Refactor mysensors light

MartinHjelmare opened the pull request:
- Add a light entity class per V_LIGHT, V_DIMMER, V_RGB and V_RGBW.
  Make these classes inherit each other up to MySensorsLight class.
- Map the entity classes to their S_TYPE in a dict.
- Check if an entity class map or just an entity class have been passed
  to pf_callback_factory before using the entity_class variable in
  homeassistant/components/mysensors.py.
- Add rgb_hex_to_list function in homeassistant/util/color.py.

MartinHjelmare committed:
Refactor mysensors light

* Add a light entity class per V_LIGHT, V_DIMMER, V_RGB and V_RGBW.
    Make these classes inherit each other up to MySensorsLight class.
* Map the entity classes to their S_TYPE in a dict.
* Check if an entity class map or just an entity class have been passed
    to pf_callback_factory before using the entity_class variable in
    homeassistant/components/mysensors.py.
* Add rgb_hex_to_list function in homeassistant/util/color.py.

balloob commented on homeassistant/components/light/mysensors.py:
```diff
@@ -109,118 +100,258 @@ def device_state_attributes(self):
+            self._values[value_type] = value
+
+
+class MySensorsLightLight(MySensorsLight):
```
A Light Light? :hamster: 

![cache-invalidation](https://cloud.githubusercontent.com/assets/1444314/13031820/3ef2ff40-d28e-11e5-8a3a-b7e3ae102c60.png)

balloob commented on homeassistant/components/light/mysensors.py:
```diff
@@ -109,118 +100,258 @@ def device_state_attributes(self):
+        if self.gateway.optimistic:
+            # optimistically assume that light has changed state
+            self._state = True
+        super().turn_on(**kwargs)
```
I don't like how super is being used here. Why not extract a method called `optimistic_update` that will only update if necessary (ie `gateway.optimistic == True`) and call this from `turn_on` and `turn_off`

balloob commented on homeassistant/components/light/mysensors.py:
```diff
@@ -109,118 +100,258 @@ def device_state_attributes(self):
+        """Turn the device off."""
+        value_type = kwargs.get('value_type')
+        value = kwargs.get('value')
+        if value_type is not None and value is not None:
```
I would update the code to be like this:
```python
if value_type is None or value is None:
    _LOGGER.warning(…)
    return
```
Will allow you to get rid of the `else`

balloob commented on homeassistant/components/light/mysensors.py:
```diff
@@ -109,118 +100,258 @@ def device_state_attributes(self):
+    def turn_off(self, **kwargs):
+        """Turn the device off."""
+        set_req = self.gateway.const.SetReq
+        value_type = kwargs.get('value_type')
```
These values will never come from `kwargs` as this method is called from the component.

balloob commented on homeassistant/components/light/mysensors.py:
```diff
@@ -109,118 +100,258 @@ def device_state_attributes(self):
+            set_req.V_DIMMER
+            if set_req.V_DIMMER in self._values else value_type)
+        value = 0 if set_req.V_DIMMER in self._values else value
+        super().turn_off(value_type=value_type, value=value)
```
I do not feel like this is the correct way of using `super`. I would suggest you rename `turn_off` in the base class to be `_turn_off` and have it be a helper method to this turn off (because that's what it is).

MartinHjelmare replied:
Just so we're on the same page, the inheritance hierarchy is like this:
```yaml
Entity:
  ToggleEntity:
    Light:
      MySensorsLight:
        MySensorsLightLight:
          MySensorsLightDimmer:
            MySensorsLightRGB:
            MySensorsLightRGBW:
```
I saw that the class docstrings are not correct for the light child classes down the line. So each update/turn_on/turn_off will traverse up the hierarchy using the super calls.

I can break out the top ```turn_off``` method in MySensorsLight into ```_turn_off```, but the ones below have to keep their names to not break the flow.

balloob replied:
You're right, I did not see the correct inheritance tree.

I'm not sure if the inheritance is making the code more clear or actually more confusing. Since you're the maintainer of this piece of code I'll leave it up to you.

MartinHjelmare replied:
After some more thought, I think you have a point, and also a good idea how this can be changed. I've basically already done it, but I then I overcomplicated it with the deep nested structure.

So... Make it one level inheritance, MySensorsLight parent with four children, one per light type. Break out one helper method per child and per update/turn_on/turn_off, and put these in the parent. Call the helper methods as needed from the child methods update/turn_on/turn_off.

balloob commented on homeassistant/util/color.py:
```diff
@@ -1,5 +1,6 @@
 """
-homeassistant.util.color
+homeassistant.util.color.
```
If you're going to follow pep257, it should be a description instead of package name.

balloob commented on homeassistant/util/color.py:
```diff
@@ -88,3 +87,11 @@ def color_xy_brightness_to_RGB(vX, vY, brightness):
     return (r, g, b)
+
+
+def rgb_hex_to_list(hex_string):
```
This will need some tests to make sure it works. I also think the name could be more descriptive, what about  `hex_to_rgb` ?

MartinHjelmare replied:
Ok. What about ```rgb_hex_to_rgb_list```?

balloob replied:
Good too

On Sun, Feb 14, 2016, 18:11 Martin Hjelmare <notifications@github.com>
wrote:

> In homeassistant/util/color.py
> <https://github.com/balloob/home-assistant/pull/1243#discussion_r52854389>
> :
>
> > @@ -88,3 +87,11 @@ def color_xy_brightness_to_RGB(vX, vY, brightness):
> >      r, g, b = map(lambda x: int(x * 255), [r, g, b])
> >
> >      return (r, g, b)
> > +
> > +
> > +def rgb_hex_to_list(hex_string):
>
> Ok. What about rgb_hex_to_rgb_list?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/balloob/home-assistant/pull/1243/files#r52854389>.
>

MartinHjelmare committed:
Change refactor structure

* Make a flatter one level inheritance, with MySensorsLight as parent
	with four children, one per light type.
* Break out helper methods. One per plain light, dimmer and RGB/RGBW
	children and per update, turn_on and turn_off, nine in total. Put
	these in the parent.
* Call the helper methods as needed from the child methods update,
	turn_on and turn_off.
* Change name of MySensorsLightLight to MySensorsLightPlain.
* Fix module docstrings according to pep257.
* Change name of color util method from rgb_hex_to_list to
	rgb_hex_to_rgb_list.
* Add unit tests for rgb_hex_to_rgb_list.

MartinHjelmare commented:
Don't merge yet, need to do live tests.

balloob commented on homeassistant/components/light/mysensors.py:
```diff
@@ -109,118 +117,268 @@ def device_state_attributes(self):
+
+    def update(self):
+        """Update the controller with the latest value from a sensor."""
+        super()._update()
```
You should only use `super()` if you're calling a method on the parent that you overwrote. Since you didn't overwrite this method, you can call `self._update()`

MartinHjelmare replied:
Yes, of course.

balloob commented on homeassistant/components/light/mysensors.py:
```diff
@@ -109,118 +117,268 @@ def device_state_attributes(self):
+
+    def __init__(self, *args):
+        """Setup instance attributes."""
+        super().__init__(*args)
```
If you're just going to call `super` you don't have to overwrite it.

MartinHjelmare replied:
Yep.

balloob commented on homeassistant/components/light/mysensors.py:
```diff
@@ -109,118 +117,268 @@ def device_state_attributes(self):
-            self._brightness = brightness
             self._white = white
+
+    def _turn_on(self):
```
This method needs either better naming or should just be inlined in the other methods.

MartinHjelmare replied:
Yes.

MartinHjelmare committed:
Clean up and fix

* Add check if V_LIGHT is in values before sending message in
	_turn_on_light.
* Replace super calls with self.
* Remove not needed init method in child classes.
* Remove turn_on method in parent class and add update_ha_state
	to _turn_on_light, _turn_on_dimmer and _turn_on_rgb_or_w.

MartinHjelmare commented:
Ok, I think I've addressed your comments, @balloob , and I've tested it live.

balloob commented:
Woop woop :dolphin: :dollar:
